### PR TITLE
Commissioning time sync: copy parameters to autocommissioner

### PR DESCRIPTION
**Pull request by cecille**
The auto commissioner is meant to take ownership of the supplied buffers from the caller. PR adds static buffers for time zone and DST. Unfortunately, neither is trivially destructible until we move from chip::Optional to std::optional. For now, they are static, we can move to dynamic allocation if required once we have support and can use proper memory management tools.

### CODE REVIEW

File: src/controller/AutoCommissioner.cpp [hunk 1]
```diff
@@ -174,6 +176,35 @@ CHIP_ERROR AutoCommissioner::SetCommissioningParameters(const CommissioningParam
     }
     mParams.SetCSRNonce(ByteSpan(mCSRNonce, sizeof(mCSRNonce)));
 
+    if (params.GetDSTOffsets().HasValue())
+    {
+        ChipLogProgress(Controller, "Setting DST offsets from parameters");
+        size_t size = params.GetDSTOffsets().Value().size();
```
Comment (andy31415) [inline src/controller/AutoCommissioner.cpp:182]: nit: consider:

```cpp
const size_t size = std::min(params.GetDSTOffsets().Value().size(), kMaxSupportedDstStructs);
```

For shorter code

**Review by andy31415 — APPROVED** [review]

File: src/controller/AutoCommissioner.cpp (hunk 1 above)
Comment (andy31415) [reply src/controller/AutoCommissioner.cpp:182]: I even wonder about:

```cpp
const size_t size = std::min(params.GetDSTOffsets().Value().size(), kMaxSupportedDstStructs);
auto list = app::DataModel::List<app::Clusters::TimeSynchronization::Structs::DSTOffsetStruct::Type>(mDstOffsetsBuf, size);

CopySpanToMutableSpan(params.GetDSTOffsets().Value().SubSpan(0, size), list);

mParams.SetDSTOffsets(list);
```

not sure if more readable though :/

Comment (cecille) [reply src/controller/AutoCommissioner.cpp]: Yes, min certainly is the correct function here, and I somehow managed to forget it existed. Which is just ... wow.

Weirdly, CopySpanToMutableSpan is only available for byte and char spans, so I can't use that here. I'm not sure WHY we don't have a templated version, but we don't.

**Review by woody-apple — APPROVED** [review]

File: src/controller/AutoCommissioner.cpp [hunk 2]
```diff
@@ -174,6 +176,33 @@ CHIP_ERROR AutoCommissioner::SetCommissioningParameters(const CommissioningParam
...
+    {
+        ChipLogProgress(Controller, "Setting Time Zone from parameters");
+        size_t size = std::min(params.GetTimeZone().Value().size(), kMaxSupportedTimeZones);
+        for (size_t i = 0; i < size; ++i)
+        {
+            mTimeZoneBuf[i] = params.GetTimeZone().Value()[i];
+            if (mTimeZoneBuf[i].name.HasValue())
+            {
+                auto span = MutableCharSpan(mTimeZoneNames[i], kMaxTimeZoneNameLen);
+                CopyCharSpanToMutableCharSpan(mTimeZoneBuf[i].name.Value(), span);
```
Comment (bzbarsky-apple) [inline src/controller/AutoCommissioner.cpp:200]: If this fails, won't this possibly leave garbage data in `mTimeZoneNames[i]`?

Comment (cecille) [reply src/controller/AutoCommissioner.cpp:200]: why would that ever fail? It's a statically allocated buffer that's been size checked.

Comment (bzbarsky-apple) [reply src/controller/AutoCommissioner.cpp:200]: I don't think `mTimeZoneBuf[i].name.Value()` has been size-checked...

**Pull request merged**